File: algorithms/cpp/twoSum/twoSum.cpp
// Source : https://leetcode.com/problems/two-sum/
// Author : Zhonghuan Gao
// Date   : 2017-11-21

/********************************************************************************** 
* 
* Given an array of integers, return indices of the two numbers such that they add up to a specific target.
*
* You may assume that each input would have exactly one solution, and you may not use the same element twice.
* 
* Example:
* Given nums = [2, 7, 11, 15], target = 9,
* Because nums[0] + nums[1] = 2 + 7 = 9,
* return [0, 1].
* 
*               
**********************************************************************************/


class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
    	// nest loop
        vector<int> result;
        int length = nums.size();
        for (int i = 0; i < length - 1; i++) {
            for (int j = i + 1; j < length; j++) {
                if (nums[i] + nums[j] == target) {
                    result.push_back(i);
                    result.push_back(j);
                    return result;
                }
            }
        }
        return result;
    }
};

File: algorithms/go/0001.twoSum/twoSum.go
package leetcode

/**
 * 解法一：
 * 暴力法，时间复杂度O(n*n)
 */
func twoSumV1(nums []int, target int) []int {
	var res []int
	for i := 0; i < len(nums); i++ {
		for j := i + 1; j < len(nums); j++ {
			if nums[i]+nums[j] == target {
				return append(res, i, j)
			}
		}
	}
	return res
}

/**
 * 解法二：
 * 遍历数组，将元素依次存入map，直到能在map中能找到对于目标值来说当前元素等另一半数
 * 时间复杂度O(n)
 */
func twoSum(nums []int, target int) []int {
	m := make(map[int]int)
	for i := 0; i < len(nums); i++ {
		another := target - nums[i]
		if _, ok := m[another]; ok {
			return []int{m[another], i}
		}
		m[nums[i]] = i
	}
	return nil
}
